bluesky_queueserver/manager/profile_ops.py

```diff
@@ -556,17 +559,189 @@ def load_list_of_plans_and_devices(path_to_file=None):
+                    },
+                    "allowed_devices": {
+                        "type": "array",
+                        "additionalItems": {"type": "string"},
+                        "items": [{"type": ["string", "null"]}],
+                    },
+                    "forbidden_devices": {
+                        "type": "array",
+                        "additionalItems": {"type": "string"},
+                        "items": [{"type": ["string", "null"]}],
+                    },
+                },
+            },
+        },
+    },
+}
+
+
+def load_user_group_permissions(path_to_file=None):
+
+    if not path_to_file:
+        return {}
+
+    try:
+
+        if not os.path.isfile(path_to_file):
+            raise IOError(f"File '{path_to_file}' does not exist.")
+
+        with open(path_to_file, "r") as stream:
+            user_group_permissions = yaml.safe_load(stream)
+
+        # if "user_groups" not in user_group_permissions:
+        #    raise IOError(f"Incorrect format of user group permissions: key 'user_groups' is missing.")
+        jsonschema.validate(instance=user_group_permissions, schema=_user_group_permission_schema)
+
+        if "root" not in user_group_permissions["user_groups"]:
+            raise Exception("Missing required user group: 'root'")
+        if "admin" not in user_group_permissions["user_groups"]:
+            raise Exception("Missing required user group: 'admin'")
+
+    except Exception as ex:
+        msg = f"Error while loading user group permissions from file '{path_to_file}': {str(ex)}"
+        raise IOError(msg)
+
+    return user_group_permissions
+
+
+def _select_allowed_items(item_dict, allow_patterns, disallow_patterns):
+    """
+    Creates the dictionary of items selected from `item_dict` such that item names
+    satisfy re patterns in `allow_patterns` and not satisfy patterns in `disallow_patterns`.
+    The function does not modify the dictionary, but creates a new dictionary with
+    selected items, where item values are deep copies of the values of the original
+    dictionary.
+
+    Parameters
+    ----------
+    item_dict: dict
+        Dictionary of items.
+    allow_patterns: list(str)
+        Selected item should match at least one of the re pattern. If it the value is ``[None]``
+        then all items are selected. If ``[]``, then no items are selected.
+    disallow_patterns: list(str)
+        Selected item should not match any of the re patterns. If the value is ``[None]``
+        or ``[]`` then no items are deselected.
+
+    Returns
+    -------
+    dict
+        Dictionary of the selected items.
+    """
+    items_selected = {}
+    for item in item_dict:
+        select_item = False
+        if allow_patterns:
+            if allow_patterns[0] is None:
+                select_item = True
+            else:
+                for pattern in allow_patterns:
+                    if re.search(pattern, item):
+                        select_item = True
```

Comment:
Should we break the for-loop here in case it's found?

Reply:
Yes, thanks. Done.